lib/api/objectCopy.js

```diff
@@ -0,0 +1,304 @@
+import async from 'async';
+import { errors } from 'arsenal';
+
+import data from '../data/wrapper';
+import kms from '../kms/wrapper';
+import { logger } from '../utilities/logger';
+import services from '../services';
+import utils from '../utils';
+import validateHeaders from '../utilities/validateHeaders';
+
+/**
+ * Preps metadata to be saved (based on copy or replace request header)
+ * @param {object} sourceObjMD - object md of source object
+ * @param {object} headers - request headers
+ * @param {boolean} sourceIsDestination - whether or not source is same as
+ * destination
+ * @param {AuthInfo} authInfo - authInfo from Vault
+ * @param {string} objectKey - destination key name
+ * @param {object} log - logger object
+ * @return {object} storeMetadataParams or an error
+ */
+function _prepMetadata(sourceObjMD, headers, sourceIsDestination, authInfo,
+    objectKey, log) {
+    let whichMetadata = headers['x-amz-metadata-directive'];
+    // Default is COPY
+    whichMetadata = whichMetadata === undefined ? 'COPY' : whichMetadata;
+    if (whichMetadata !== 'COPY' && whichMetadata !== 'REPLACE') {
+        return { error: errors.InvalidArgument };
+    }
+    const overrideMetadata = {};
+    if (headers['x-amz-server-side-encryption']) {
+        overrideMetadata['x-amz-server-side-encryption'] =
+            headers['x-amz-server-side-encryption'];
+    }
+    if (headers['x-amz-storage-class']) {
+        overrideMetadata['x-amz-storage-class'] =
+            headers['x-amz-storage-class'];
+    }
+    if (headers['x-amz-website-redirect-location']) {
+        overrideMetadata['x-amz-website-redirect-location'] =
+            headers['x-amz-website-redirect-location'];
+    }
+    // Cannot copy from same source and destination if no MD
+    // changed
+    if (sourceIsDestination && whichMetadata === 'COPY' &&
+        Object.keys(overrideMetadata).length === 0) {
+        return { error: errors.InvalidRequest };
```

Comment:
Let's use Arsenal's custom description here as the default desc. relates to SOAP requests. AWS is using

> This copy request is illegal because it is trying to copy an object to itself without changing the object\'s metadata, storage class, website redirect location or encryption attributes.